joomla/joomla-cms
# Admin Menu custom presets, import/export a preset into/from database driven menu

### Summary of Changes

As we know that with the release of Joomla 3.7 we have the Admin Menu Manager that lets you customise the administrator menu according to your needs. The backend menu comes in two flavours: **Custom Menu** and **Preset Menu**. You can have as many custom menu each configured in different way as suitable. Or if you do not want to customise then you use the Preset menu.

This PR is indented to extend the features for this Admin Menu Manager as promised in the initial strategy.


### Changes

1. **Custom Presets**: 
If you copy (and modify) one of the existing preset files from `/administrator/com_menus/presets` to `/administrator/templates/{your_template}/html/com_menus/presets` then you can use it as your own custom preset.

2. **Import Preset Menu**: 
If you would like to modify merely one or two things and keep rest of the menu structure the same way as in a preset menu you still need to create all of those links one by one yourself **manually**. After this PR Joomla will allow you to directly import an existing preset (core or custom) into your custom admin menu directly and then you can customise it very quickly. When creating/editing a `Menu` (menutype) you can optionally select a preset that you would like to import.

3. **Export Menu as Preset**: 
You can also export your database driven `custom admin menu` as an a Preset. Just select a Menu in the list and click `Download As Preset` button in the toolbar.

4. **Menu Module Layout Override**: 
The current implementation of the Admin Menu Module has the HTML rendering from within the PHP class `JAdminCssMenu` which does not let you customise the HTML in any way. This PR puts all the HTML output logic into the layout files under `tmpl/` folder as commonly expected. This way you can easily override the layout in your template.

5. **Menu Structure Override**: 
Prior to Joomla 3.7 we could override the menu links using `default_enabled.php` and `default_disabled.php`. After the Menu Manager was introduced this ability was removed. Most likely you would really never need this anymore. It is always better to import a preset in a custom menu and customise it in the Menu Manager instead of messing with any PHP code. However, if you still insist, you can use the **Custom Preset** as discussed above. The naming convention is like `my-custom-menu.xml` becomes `My Custom Menu` (we can improve this later if needed). It is really easy to understand and work with. Give it a try.


### Testing Instructions
1. Create a custom admin menu and set it as your active admin menu in the module settings.
    * It should work correct and show the links based on the appropriate permissions.
    * Notice about missing important menu links (if any) should be displayed as before.
    * Clicking on the Menu Recovery link should enable the recovery mode and you should be able to exit the recovery mode too.
    * Components container should also work correctly like hiding/showing choices as configured.
    * Installing a new component should make the link under the components container menu.
 
1. Try to choose another preset in the module settings. The displayed menu should be updated according the selected preset.
 1. Create a custom preset by copying (as discussed above) and make some changes. Set this custom preset as your active menu. The menu should now load you custom preset.
 1. Create new custom admin menu and choose a preset to import. This should load the menu links and hierarchy from the preset.
 1. From the list view of Menu, select one of your Admin Menus and click on "Download as Preset" button. The XML file should be downloaded correctly and should contain the selected menu links with correct settings.


### Documentation Changes Required
I'd like to write the detailed documentation of the entire Admin Menu Manager. Please advise where should I head towards.

---
PS: As I am on a vacation this PR was created in hurry. Pardon me for any mistakes and please ask for any kind of clarifications or suggestions. I'll try to respond as quickly as possible.
